workflows/pipe-common/scripts/autoscale_sge.py

```diff
@@ -1300,10 +1311,118 @@ def _validate_existence(self, host, hosts_dict):
+        last_monitored_timestamps.last_tag_timestamp = self.clock.now()
+        self.last_monitored_hosts.update({host: last_monitored_timestamps})
+
+    def _untag_run(self, host, timestamp=None, last_monitored_timestamps=None):
+        run_id = self.common_utils.get_run_id_from_host(host)
+        self._remove_worker_tag(run_id)
+        if not last_monitored_timestamps:
+            del self.last_monitored_hosts[host]
+            return
+        last_monitored_timestamps.last_action_timestamp = timestamp
+        last_monitored_timestamps.last_tag_timestamp = None
+        self.last_monitored_hosts.update({host: last_monitored_timestamps})
+
+    def _process_current_hosts(self, current_hosts, hosts_activity):
+        Logger.info("Started tags processing for running hosts '%s' ." % current_hosts)
+        for current_host in current_hosts:
+            timestamp = hosts_activity[current_host]
+            if not timestamp:
+                continue
+            last_monitored_timestamps = self.last_monitored_hosts.get(current_host, None)
+            if not last_monitored_timestamps:
+                Logger.info("Found not monitored host '%s'." % current_host)
+                last_action_marker = LastActionMarker()
+                self.last_monitored_hosts.update({current_host: last_action_marker})
+                continue
+            if self._run_is_active(timestamp):
+                if not last_monitored_timestamps.last_tag_timestamp:
+                    Logger.info("Host '%s' is active. Adding a new tag to the corresponding run." % current_host)
+                    self._tag_run(current_host, timestamp, last_monitored_timestamps)
+                    Logger.info("Added tag to run for host '%s'." % current_host)
+                # do nothing if active run was already tagged
+                continue
+            Logger.info("Host '%s' is not active." % current_host)
+            if last_monitored_timestamps and last_monitored_timestamps.last_tag_timestamp:
+                Logger.info("Removing tag from run.")
+                self._untag_run(current_host, timestamp, last_monitored_timestamps)
+                Logger.info("Removed tag from run for host %s." % current_host)
+
+    def _process_outdated_hosts(self, monitored_hosts, current_hosts):
+        for monitored_host in monitored_hosts:
+            if monitored_host not in current_hosts:
+                Logger.info("Host %s scaled down. Removing tags from run." % monitored_host)
+                self._untag_run(monitored_host)
+
+    def _add_worker_tag(self, run_id):
+        run = self.api.load_run(run_id)
+        tags = run.get('tags') or {}
+        tags.update({self._WORKER_TAG: 'true'})
+        self.api.update_pipeline_run_tags(run_id, tags)
+
+    def _remove_worker_tag(self, run_id):
+        run = self.api.load_run(run_id)
+        tags = run.get('tags') or {}
+        if tags.__contains__(self._WORKER_TAG):
```

Comment:
Let's use more pythonish syntax:

```suggestion
        if self._WORKER_TAG in tags:
```